src/qibolab/_core/instruments/keysight/qcs.py

```diff
@@ -0,0 +1,172 @@
+"""Qibolab driver for Keysight QCS instrument set."""
+
+from collections import defaultdict
+from functools import reduce
+from typing import ClassVar, Optional
+
+import numpy as np
+from keysight import qcs
+
+from qibolab._core.components import AcquisitionChannel, Config, DcChannel, IqChannel
+from qibolab._core.execution_parameters import AveragingMode, ExecutionParameters
+from qibolab._core.identifier import ChannelId, Result
+from qibolab._core.instruments.abstract import Controller
+from qibolab._core.pulses import PulseId
+from qibolab._core.sequence import InputOps, PulseSequence
+from qibolab._core.sweeper import ParallelSweepers
+from qibolab._core.unrolling import Bounds
+
+from .pulse import (
+    process_acquisition_channel_pulses,
+    process_dc_channel_pulses,
+    process_iq_channel_pulses,
+)
+from .results import fetch_result, parse_result
+from .sweep import process_sweepers
+
+NS_TO_S = 1e-9
+
+BOUNDS = Bounds(
+    waveforms=1,
+    readout=1,
+    instructions=1,
+)
+
+__all__ = ["KeysightQCS"]
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
+    # Map of Qibolab channel IDs to QCS virtual channels
+    virtual_channel_map: dict[ChannelId, qcs.Channels]
```

Comment:
Looking at the example platform here(*): https://github.com/qiboteam/qibolab_platforms_nqch/blob/qibolab-0.2/iqm5q/platform.py, I have the impression that this dictionary can be eliminated by using qibolab's `channel.path`, which I believe is currently not used. I am not sure if types can be worked out though, because our `channel.path` should be a `str`. If `qcs.ChannelMapper` provides string keys, you may be able to retain the information using just `qcs_channel_map` and `channel.path`, otherwise it may be tricky.

In any case, I have no experience with these particular instruments, so if you find it easier to use it this way, it is fine with me.

(*) Unrelated to the above, it would be useful to also have an example (not necessarily real/working) platform here, but we can postpone this for later, when we have better testing or even better documentation for how to use each driver.

Reply:
I prefer to have a direct mapping because other approaches would involve some type casting to get the necessary indices from the `channel` object.